log.go

```diff
@@ -177,6 +132,93 @@ func (el *eventLogger) Event(ctx context.Context, event string, metadata ...Logg
 	WriterGroup.Write(append(out, '\n'))
 }
 
+// EventBegin starts an EventInProgress and returns it. The eip is to be
+// completed at a later time
+// Example usage:
+//
+// func SomeFunction() (err error) {
+//    eip := log.EventBegin(ctx, "DoesSomething")
+//    defer func() {
+//        eip.DoneWithErr(err)
+//    }()
+//    ...
+//  }
+func (el *eventLogger) EventBegin(ctx context.Context, event string, metadata ...Loggable) *EventInProgress {
+	_, eip := el.eventBeginHelper(ctx, event, metadata...)
+	return eip
+}
+
+// EventBeginInContext starts an EventInProgress, stores it in the context, and
+// returns the new context. The eip can be completed at a later time using the
+// `MaybeFinishEvent()` method
+// Example usage:
+//
+// func SomeFunction(ctx) {
+//    ctx := log.EventBeginInContext(ctx, "DoesSomething")
+//    ...
+//    go func() {
+//        defer logging.MaybeFinishEvent(ctx)
+//        ...
+//    }
+//    ...
+//  }
+func (el *eventLogger) EventBeginInContext(ctx context.Context, event string, metadata ...Loggable) context.Context {
```

Comment:
I'm not the biggest fan of this name, but I'm not sure what else to call it...suggestions?

Reply:
Yeah... I kind of have to agree. However, if we're planning on transitioning everything to using this method, we could call it something *totally* different like `Span()` (and eventually deprecate and remove the other methods).

Also, I'm wondering if this should return a custom type (custom interface?). Something that combines the features of `EventInProgress` and `context.Context` such that it can be coerced into a normal `context.Context`.

```go
type Span interface {
    context.Context
    Finish()
    FinishWithErr(err)
}
```

You previously suggested making `EventInProgress` implement `Context` but that would have forced us to do a mass refactor up-front (as we would have needed to rename `Done`). By using a new method with a new type, we'd just slowly transition to the new method. Thoughts? Improvements? Alternatives?

Reply:
By "coerce" I mean that we can pass it to a function that expects a `context.Context` and go will just "do the right thing".

Reply:
I completely agree, let's continue the disunion here - https://github.com/ipfs/go-log/issues/19